DR2-1242 Add court document anonymiser.

MancunianSam opened the pull request:
This is the anonymiser script for the court document handlers.

The description of what it does is in the README.

techncl commented on tests/integration_test.rs:
```diff
@@ -0,0 +1,145 @@
+    }
+    "#
+}
+fn create_package(input_dir: &TempDir, json: &str, file_name: Option<String>) {
```
MIssing newline before

techncl commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,21 @@
+name: DR2 Release Binary
```
Should we add a more descriptive name, like `DR2 Release Package Anonymiser Binary`?

MancunianSam replied:
I'm not sure about "Package". DR2 Release Anonymiser Binary sounds ok though.

techncl reviewed (commented):
Looks good, just some minor comments, plus:

1. I was wondering if it might be worth it to add the types? In some cases, I'm not sure if something is a String or an Object by name
2. I think from Rust 1.58 onwards, instead of the `format!("TRE-{}-metadata.json", &input_batch_reference)` form, you can do `format!("TRE-{&input_batch_reference}-metadata.json")`
3. Do you think it's worth checking if the files have been removed?
4. Do you think it's worth adding `rustfmt`?
5. Testing for errors - what about testing for:
     - `file.file_name()` on line 62
     - batch reference on line 68

MancunianSam commented on install.sh:
```diff
@@ -0,0 +1,24 @@
+}
+
+download() {
+	if check_cmd curl; then
```
I edited it in vim which insists on using the largest tabs ever. I've switched it back to two spaces.

MancunianSam commented on src/main.rs:
```diff
@@ -0,0 +1,170 @@
+fn is_hidden(entry: &DirEntry) -> bool {
+    entry.file_name()
+        .to_str()
+        .map(|s| s.starts_with("."))
```
`file_name` is good.

MancunianSam commented on src/main.rs:
```diff
@@ -0,0 +1,170 @@
+    let path_list = fs::read_dir(directory_path).unwrap()
+        .filter_map(|e| {
+            let entry = e.ok()?;
+            if !is_hidden(&entry) && !is_directory(&entry) { Some(entry.path()) } else { None }
```
Yep, makes sense. I've swapped them around in the if and changed them to `is_file` and `is_not_hidden`

MancunianSam commented on src/main.rs:
```diff
@@ -0,0 +1,170 @@
+}
+
+fn update_json_file(metadata_file_name: &PathBuf, checksum: String, json_value: &mut Value) -> Result<(), Error> {
+    json_value["parameters"]["TDR"]["Contact-Email"] = json!("XXXXXXXXX");
```
You can, I've moved it to one called `tdr`

MancunianSam commented on src/main.rs:
```diff
@@ -0,0 +1,170 @@
+}
+fn parse_json(metadata_file_path: &PathBuf) -> Result<Value, Error> {
+    let mut file = File::open(metadata_file_path)?;
+    let mut data = String::new();
```
`metadata_json_as_string` is good.

MancunianSam commented on src/main.rs:
```diff
@@ -0,0 +1,170 @@
+    archive.unpack(&output_path)?;
+    Ok(())
+}
+fn parse_json(metadata_file_path: &PathBuf) -> Result<Value, Error> {
```
I'm blaming IntelliJ for the newline issues. It shows me the usage above the function so it looks like there's one there :smile: 
I've renamed the function.

MancunianSam commented on src/main.rs:
```diff
@@ -0,0 +1,170 @@
+    }
+}
+
+fn process_package(dir_output: &PathBuf, file: &PathBuf) -> Result<(), Error> {
```
1. I tried splitting it out but because there are so many input paths and output paths to various files, you end up with functions which take a lot of arguments and return a lot of arguments as a tuple. I couldn't find a way of doing it neatly so I've left it as is.
2. I'll leave it as file because we're not explicitly searching for tar files and I hope one day they'll stop zipping them up anyway so we can skip that step.

MancunianSam commented on src/main.rs:
```diff
@@ -0,0 +1,170 @@
+    fs::rename(&extracted_output_original_name, &extracted_output_path)?;
+    fs::rename(&metadata_input_file_path, &metadata_output_file_path)?;
+
+    let mut json_value = parse_json(&metadata_output_file_path)?;
```
It's not a file, it's a json `Value` object which is clearer now I've added the types in. I've gone with `metadata_json_value`

MancunianSam commented on src/main.rs:
```diff
@@ -0,0 +1,170 @@
+
+    create_docx(&docx_path_string, judgment_name)?;
+
+    let checksum = checksum(&docx_path_string);
```
`docx_checksum` works.

MancunianSam commented on tests/integration_test.rs:
```diff
@@ -0,0 +1,145 @@
+use assert_fs::TempDir;
+use serde_json::Value;
+
+fn missing_filename() -> &'static str {
```
`json_missing_filename` is good.

MancunianSam commented on src/main.rs:
```diff
@@ -0,0 +1,170 @@
+
+    let output_tar_gz_path: PathBuf = Path::new(&dir_output).join(Path::new(&tar_gz_file_name.replace("TDR", "TST")));
+    let uncompressed_folder_input_path = &file.with_extension("").with_extension("");
+    let input_batch_reference = uncompressed_folder_input_path.file_name().and_then(|name| name.to_str())
```
I added `rustfmt` and it did this exact thing so yes :smile:

MancunianSam commented:
Thanks for looking at this.

1. Definitely worth adding the types, I've put them in the main.rs file.
2. I didn't know that, that's much neater. I've changed it, thanks.
3. The reason I wasn't is because they're not guaranteed to be there but I've put in a new function which checks if the path exists and then deletes them. The result is kept now so if remove_file returns an error, it will show in the console.
4. Yep, I've added it in
5. I don't think it's necessary. The reason it returns errors is because of some unlikely encoding thing that's possible with the file paths. Considering this is only going to be used on *nix systems with known style filenames, I think it's ok as is.

techncl commented on src/main.rs:
```diff
@@ -56,72 +48,107 @@ fn main() {
-    let judgment_name = json_value["parameters"]["PARSER"]["name"].to_string();
-    let docx_path_string = extracted_output_path.join(PathBuf::from(docx_file_name));
+    let judgment_name: String = metadata_json_value["parameters"]["PARSER"]["name"].to_string();
+    let docx_path_string: PathBuf = extracted_output_path.join(PathBuf::from(docx_file_name));
```
Should this be just `docx_path`?

MancunianSam replied:
It should, good spot. It used to be a string before I realised using PathBuf was better.

techncl reviewed (commented):
Just one minor comment
